### DISCUSSION TITLE
How do I send websocket message from the client?

### DISCUSSION THREAD

**Discussion by dulguun0225** [enisdenjo/graphql-ws#294] [Q&A] [answered]
Hello. I'm migrating a existing project to `graphql-ws`. There's a code that sends websocket message from the client like 
```javascript
    wsLink.subscriptionClient.sendMessage({
      type: 'messengerConnected',
      value: messengerData
    });
```
I want to send a websocket message from the client just like the previous code. How do I do that?

**Comment by enisdenjo** [comment, marked as answer]
You cannot. `graphql-ws` is in control over the message transport. You can only `subscribe`.

Even if you could send a custom message (like the one from your example), a GraphQL over WebSocket compliant server would terminate the connection immediately since [invalid messages are considered fatal](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#invalid-message).

**Reply by dulguun0225** [reply]
Well, then, I'm in a pickle now. Any advice would be greatly appreciated. 

Basically what I'm trying to do is, 
1. Send some data from client to server
2. Attach that information to the corresponding server subscription socket. 
3. Use that information on subscription's `onClose` event.

**Reply by enisdenjo** [reply]
As already mentioned, [any message you send that's not a part of the GraphQL over WebSocket protocol is considered invalid and therefore fatal](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#message-types).

You can instead use the [client's `connectionParams`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#connectionparams), which are a part of the [`ConnectionInit` message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#connectioninit), to transmit additional data about the connection.

**Reply by dulguun0225** [reply]
Thank you!
